cedar-lean/Cedar/Partial/Response.lean

```diff
@@ -0,0 +1,171 @@
+/-
+ Copyright Cedar Contributors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Data.Set
+import Cedar.Partial.Evaluator
+import Cedar.Partial.Expr
+import Cedar.Spec.Policy
+
+/-!
+This file defines Cedar partial responses.
+-/
+
+namespace Cedar.Partial
+
+open Cedar.Data
+open Cedar.Spec (Effect Error PolicyID)
+open Cedar.Spec.Effect
+
+inductive Residual where
+  | residual (id : PolicyID) (effect : Effect) (condition : Partial.Expr)
+  | error (id : PolicyID) (error : Error) -- definitely results in this error, for any substitution of the unknowns
+
+deriving instance Repr, DecidableEq, Inhabited for Residual
+
+def Residual.id : Residual → PolicyID
+  | .residual id _ _ => id
+  | .error id _ => id
+
+def Residual.effect : Residual → Option Effect
+  | .residual _ effect _ => effect
```

Comment:
Ha. Turns out Lean has a built-in impl of Coe for T to Option T that applies `some`.  Should we take advantage of this (as here) or adopt a style guidance that we always explicitly state `some`?

Reply:
Ah, cool. I prefer the `some` (since I'm used to having to do that in other languages), but I don't feel strongly. Perhaps @emina has stronger feelings.

Reply:
FTR I do not have a strong feeling on this and don't remember explicitly deciding to omit the `some` here, I think it was unintentional.  It's certainly the case that elsewhere in this repo we use the explicit `some` in most (all?) places